src/crypto/sign.js

```diff
@@ -0,0 +1,298 @@
+	const encryptedMessage = bufferToHex(cipherBytes);
+
+	return {
+		nonce: nonceHex,
+		encryptedMessage,
+	};
+}
+
+/**
+ * @method decryptMessageWithSecret
+ * @param cipherHex
+ * @param nonce
+ * @param secret
+ * @param senderPublicKey
+ *
+ * @return {string}
+ */
+
+export function decryptMessageWithSecret(cipherHex, nonce, secret, senderPublicKey) {
+	const recipientPrivateKeyBytes = getRawPrivateAndPublicKeyFromSecret(secret).privateKey;
+	const convertedPrivateKey = convertPrivateKeyEd2Curve(recipientPrivateKeyBytes);
+	const senderPublicKeyBytes = hexToBuffer(senderPublicKey);
+	const convertedPublicKey = convertPublicKeyEd2Curve(senderPublicKeyBytes);
+	const cipherBytes = hexToBuffer(cipherHex);
+	const nonceBytes = hexToBuffer(nonce);
+
+	const decoded = naclInstance.crypto_box_open(
+		cipherBytes, nonceBytes, convertedPublicKey, convertedPrivateKey,
+	);
+
+	return naclInstance.decode_utf8(decoded);
+}
+
+/**
+ * @method signTransaction
+ * @param transaction Object
+ * @param givenKeys Object
+ *
+ * @return {string}
+ */
+
+export function signTransaction(transaction, givenKeys) {
+	const transactionHash = getTransactionHash(transaction);
+	const signature = naclInstance.crypto_sign_detached(transactionHash, Buffer.from(givenKeys.privateKey, 'hex'));
+	return Buffer.from(signature).toString('hex');
+}
+
+/**
+ * @method multiSignTransaction
+ * @param transaction Object
+ * @param givenKeys Object
+ *
+ * @return {string}
+ */
+
+export function multiSignTransaction(transaction, givenKeys) {
+	const transactionToSign = Object.assign({}, transaction);
+	delete transactionToSign.signature;
+	delete transactionToSign.signSignature;
+	const { privateKey } = givenKeys;
+	const bytes = getTransactionBytes(transactionToSign);
+	const transactionHash = getSha256Hash(bytes);
+	const signature = naclInstance.crypto_sign_detached(
+		transactionHash, hexToBuffer(privateKey),
+	);
+
+	return Buffer.from(signature).toString('hex');
+}
+
+/**
+ * @method verifyTransaction
+ * @param transaction Object
+ * @param secondPublicKey
+ *
+ * @return {boolean}
+ */
+
+export function verifyTransaction(transaction, secondPublicKey) {
+	let remove = transaction.signSignature ? 128 : 64;
+	if (secondPublicKey) remove = 64;
+	const transactionBytes = getTransactionBytes(transaction);
+	const transactionBytesWithoutSignature = Buffer
+		.alloc(transactionBytes.length - remove)
+		.fill(transactionBytes);
+	const transactionHash = getSha256Hash(transactionBytesWithoutSignature);
+
+	const signatureBuffer = transaction.signSignature
+		? Buffer.from(transaction.signSignature, 'hex')
+		: Buffer.from(transaction.signature, 'hex');
+	const senderPublicKeyBuffer = transaction.signSignature
+		? Buffer.from(secondPublicKey, 'hex')
+		: Buffer.from(transaction.senderPublicKey, 'hex');
+	const verification = naclInstance.crypto_sign_verify_detached(
+		signatureBuffer, transactionHash, senderPublicKeyBuffer,
+	);
+
+	return verification;
```

Comment:
Well now we're not verifying the first signature if the second is provided, right?